src/test/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloaderTest.java

```diff
@@ -321,6 +323,123 @@ public void downloadFrom2UrlsBothSocketTimeoutDuringBodyRead()
     }
   }
 
+  @Test
+  public void downloadFromMirroredUrlOk() throws IOException, InterruptedException {
+    try {
+      try (ServerSocket server = new ServerSocket(0, 1, InetAddress.getByName(null))) {
+        @SuppressWarnings("unused")
+        Future<?> possiblyIgnoredError =
+            executor.submit(
+                () -> {
+                  try (Socket socket = server.accept()) {
+                    Map<String, String> m = new HashMap<String, String>();
+                    readHttpRequest(socket.getInputStream(),m);
+                    String reqUri = m.getOrDefault("x-request-uri","");
+                    sendLines(
+                        socket,
+                        "HTTP/1.1 200 OK",
+                        "Date: Fri, 31 Dec 1999 23:59:59 GMT",
+                        "Connection: close",
+                        "Content-Type: text/plain",
+                        "",
+                        reqUri);
+                  }
+                  return null;
+                });
+
+        // set mirror
+        httpDownloader.setMirror(new URL(String.format("http://localhost:%d/", server.getLocalPort())));
+
+        Path resultingFile =
+            downloadManager.download(
+                Collections.singletonList(
+                    new URL(String.format("http://localhost:%d/foo", server.getLocalPort()))),
+                Collections.emptyMap(),
+                Optional.absent(),
+                "testCanonicalId",
+                Optional.absent(),
+                fs.getPath(workingDir.newFile().getAbsolutePath()),
+                eventHandler,
+                Collections.emptyMap(),
+                "testRepo");
+
+        assertThat(new String(readFile(resultingFile), UTF_8)).isEqualTo(String.format("/localhost:%d/foo", server.getLocalPort()));
+      }
+    } finally {
+      httpDownloader.setMirror(null);
+    }
+  }
+
+  @Test
+  public void downloadFromMirroredUrlFailedButOriginOk() throws IOException, InterruptedException {
+    try {
+      try (ServerSocket server1 = new ServerSocket(0, 1, InetAddress.getByName(null));
+        ServerSocket server2 = new ServerSocket(0, 1, InetAddress.getByName(null))) {
+        @SuppressWarnings("unused")
+        Future<?> possiblyIgnoredError =
+            executor.submit(
+                () -> {
+                    Socket socket = server1.accept();
+                    readHttpRequest(socket.getInputStream());
+                    sendLines(
+                        socket,
+                        "HTTP/1.1 200 OK",
+                        "Date: Fri, 31 Dec 1999 23:59:59 GMT",
+                        "Connection: close",
+                        "Content-Type: text/plain",
+                        "",
+                        "hello");
+
+                    // Never close the socket to cause SocketTimeoutException during body read on client
+                    // side.
+                    return null;
```

Comment:
Can we change this test instead to immediately return an HTTP 404, as I think this is the more common case for "the mirror does not have my file yet". :)

Reply:
Yes, you are right.  HTTP 404 is much better in this case:)

In our case we have a customized mirror server that will fetch artifacts from the original server automatically when it can't find the requested files locally. Then we don't need to upload the added/updated artifacts to our mirror server manually when WORKSPACE reviewed.